File: tools/convert_to_templates.lua
#!/usr/bin/env lua

-- Tool to convert map objects to use templates
-- Usage: lua tools/convert_to_templates.lua res/map/sandbox.tmj

local json = require('src.utils.json')
local lfs = require('lfs')

local function readFile(path)
	local file = io.open(path, 'r')
	if not file then return nil end
	local contents = file:read('*a')
	file:close()
	return contents
end

local function writeFile(path, contents)
	local file = io.open(path, 'w')
	if not file then error('Cannot write to ' .. path) end
	file:write(contents)
	file:close()
end

-- Load all templates from res/editor/
local function loadTemplates()
	local templates = {}
	for file in lfs.dir('res/editor') do
		if file:match('%.tj$') then
			local name = file:gsub('%.tj$', '')
			local contents = readFile('res/editor/' .. file)
			if contents then
				local tmpl = json.decode(contents)
				if tmpl and tmpl.object then
					templates[name] = tmpl
				end
			end
		end
	end
	return templates
end

-- Find matching template for an object
local function findTemplate(obj, templates)
	-- Try matching by type first
	for name, tmpl in pairs(templates) do
		if tmpl.object.type == obj.type then
			return name, tmpl
		end
	end
	-- Fallback: match by gid (if template has gid and object has gid)
	for name, tmpl in pairs(templates) do
		if tmpl.object.gid and obj.gid and tmpl.object.gid == obj.gid then
			return name, tmpl
		end
	end
	return nil, nil
end

-- Check if two property tables are equivalent
local function propsEqual(templateProps, objectProps)
	if not templateProps and not objectProps then return true end
	if not templateProps or not objectProps then return false end
	if #templateProps ~= #objectProps then return false end
	
	-- Build lookup from template props
	local tmplLookup = {}
	for _, p in ipairs(templateProps) do
		tmplLookup[p.name] = p
	end
	
	for _, p in ipairs(objectProps) do
		local tmplProp = tmplLookup[p.name]
		if not tmplProp then return false end
		if tmplProp.type ~= p.type then return false end
		if tmplProp.value ~= p.value then return false end
	end
	return true
end

-- Get properties that differ from template
local function getDiffProps(templateProps, objectProps)
	if not objectProps then return nil end
	if not templateProps then return objectProps end
	
	local tmplLookup = {}
	for _, p in ipairs(templateProps) do
		tmplLookup[p.name] = p
	end
	
	local diff = {}
	for _, p in ipairs(objectProps) do
		local tmplProp = tmplLookup[p.name]
		if not tmplProp or tmplProp.type ~= p.type or tmplProp.value ~= p.value then
			table.insert(diff, p)
		end
	end
	
	if #diff == 0 then return nil end
	return diff
end

-- Convert a map file
local function convertMap(mapPath)
	local contents = readFile(mapPath)
	if not contents then
		error('Map file not found: ' .. mapPath)
	end
	
	local map = json.decode(contents)
	if not map then
		error('Invalid JSON in map file')
	end
	
	local templates = loadTemplates()
	print('Loaded ' .. #templates .. ' templates')
	
	local converted = 0
	for _, layer in ipairs(map.layers) do
		if layer.type == 'objectgroup' and layer.objects then
			for _, obj in ipairs(layer.objects) do
				local tmplName, tmpl = findTemplate(obj, templates)
				if tmpl then
					local diffProps = getDiffProps(tmpl.object.properties, obj.properties)
					
					-- Build new object with template reference
					local newObj = {
						id = obj.id,
						template = '../editor/' .. tmplName .. '.tj',
						x = obj.x,
						y = obj.y
					}
					
					-- Only include properties that differ from template
					if diffProps then
						newObj.properties = diffProps
					end
					
					-- Preserve rotation if non-zero
					if obj.rotation and obj.rotation ~= 0 then
						newObj.rotation = obj.rotation
					end
					
					-- Preserve visible if false
					if obj.visible == false then
						newObj.visible = false
					end
					
					-- Preserve name if different from template
					if obj.name and obj.name ~= '' and obj.name ~= tmpl.object.name then
						newObj.name = obj.name
					end
					
					-- Replace object
					for k, v in pairs(newObj) do
						obj[k] = v
					end
					-- Remove keys not in newObj
					for k in pairs(obj) do
						if newObj[k] == nil then
							obj[k] = nil
						end
					end
					
					converted = converted + 1
					print('  Converted object ' .. obj.id .. ' (' .. obj.type .. ') -> ' .. tmplName)
				else
					print('  No template for object ' .. obj.id .. ' (' .. (obj.type or 'no type') .. ')')
				end
			end
		end
	end
	
	-- Write back
	writeFile(mapPath, json.encode(map))
	print('Converted ' .. converted .. ' objects. Saved to ' .. mapPath)
end

-- Main
local mapPath = arg[1]
if not mapPath then
	print('Usage: lua tools/convert_to_templates.lua <map_file.tmj>')
	os.exit(1)
end

convertMap(mapPath)

File: tools/convert_to_templates.py
#!/usr/bin/env python3

"""
Tool to convert map objects to use templates.
Usage: python tools/convert_to_templates.py res/map/sandbox.tmj
"""

import json
import os
import sys

def load_templates():
    """Load all templates from res/editor/"""
    templates = {}
    editor_dir = 'res/editor'
    for file in os.listdir(editor_dir):
        if file.endswith('.tj'):
            name = file[:-3]  # remove .tj
            path = os.path.join(editor_dir, file)
            with open(path, 'r') as f:
                tmpl = json.load(f)
                if tmpl and 'object' in tmpl:
                    templates[name] = tmpl
    return templates

def find_template(obj, templates):
    """Find matching template for an object by type or gid"""
    # Try matching by type first
    for name, tmpl in templates.items():
        if tmpl['object'].get('type') == obj.get('type'):
            return name, tmpl
    # Fallback: match by gid
    for name, tmpl in templates.items():
        if tmpl['object'].get('gid') and obj.get('gid') and tmpl['object']['gid'] == obj['gid']:
            return name, tmpl
    return None, None

def props_equal(template_props, object_props):
    """Check if two property lists are equivalent"""
    if not template_props and not object_props:
        return True
    if not template_props or not object_props:
        return False
    if len(template_props) != len(object_props):
        return False
    
    tmpl_lookup = {p['name']: p for p in template_props}
    for p in object_props:
        tmpl_prop = tmpl_lookup.get(p['name'])
        if not tmpl_prop:
            return False
        if tmpl_prop['type'] != p['type']:
            return False
        if tmpl_prop['value'] != p['value']:
            return False
    return True

def get_diff_props(template_props, object_props):
    """Get properties that differ from template"""
    if not object_props:
        return None
    if not template_props:
        return object_props
    
    tmpl_lookup = {p['name']: p for p in template_props}
    diff = []
    for p in object_props:
        tmpl_prop = tmpl_lookup.get(p['name'])
        if not tmpl_prop or tmpl_prop['type'] != p['type'] or tmpl_prop['value'] != p['value']:
            diff.append(p)
    
    return diff if diff else None

def convert_map(map_path):
    with open(map_path, 'r') as f:
        map_data = json.load(f)
    
    templates = load_templates()
    print(f'Loaded {len(templates)} templates')
    
    converted = 0
    for layer in map_data.get('layers', []):
        if layer.get('type') == 'objectgroup' and 'objects' in layer:
            for obj in layer['objects']:
                tmpl_name, tmpl = find_template(obj, templates)
                if tmpl:
                    diff_props = get_diff_props(tmpl['object'].get('properties'), obj.get('properties'))
                    
                    # Build new object with template reference
                    new_obj = {
                        'id': obj['id'],
                        'template': f'../editor/{tmpl_name}.tj',
                        'x': obj['x'],
                        'y': obj['y']
                    }
                    
                    # Only include properties that differ from template
                    if diff_props:
                        new_obj['properties'] = diff_props
                    
                    # Preserve rotation if non-zero
                    if obj.get('rotation', 0) != 0:
                        new_obj['rotation'] = obj['rotation']
                    
                    # Preserve visible if false
                    if obj.get('visible') is False:
                        new_obj['visible'] = False
                    
                    # Preserve name if different from template
                    if obj.get('name') and obj['name'] != '' and obj['name'] != tmpl['object'].get('name', ''):
                        new_obj['name'] = obj['name']
                    
                    # Replace object in place
                    obj.clear()
                    obj.update(new_obj)
                    
                    converted += 1
                    print(f'  Converted object {obj["id"]} ({obj.get("type", "no type")}) -> {tmpl_name}')
                else:
                    print(f'  No template for object {obj["id"]} ({obj.get("type", "no type")})')
    
    # Write back with pretty formatting
    with open(map_path, 'w') as f:
        json.dump(map_data, f, indent=2)
    
    print(f'Converted {converted} objects. Saved to {map_path}')

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Usage: python tools/convert_to_templates.py <map_file.tmj>')
        sys.exit(1)
    
    convert_map(sys.argv[1])